File: lib/api/list.js
var _ = require('lodash'),
    Promise = require('bluebird'),
    util = require('../util'),
    log = util.logger();

var LISTS = util.createPath('lists.json');

function List(marketo, connection) {
  this._marketo = marketo;
  this._connection = connection;
}

List.prototype = {
  find: function(options) {
    var arrayFields = ['id', 'name', 'programName', 'workspaceName'];
    options = _.extend({}, options, {
      _method: 'GET'
    });
    options = util.arrayToCSV(options, arrayFields);
    return this._connection.post(LISTS, {data: options});
  },

  addLeadsToList: function(listId, input) {
    if (!_.isArray(input) && !_.isEmpty(input)) {
      var msg = 'input must be an array of lead ids';
      log.error(msg);
      return Promise.reject(msg);
    }

    var path = util.createPath('lists', listId, 'leads.json');

    input = _.compact(_.map(input, function(id) {
      if (_.isString(id) || _.isNumber(id)) {
        return {id: id};
      }

      if (_.isObject(id) && _.has(id, 'id')) {
        return _.pick(id, 'id');
      }
    }));

    return this._connection.postJson(path, {input: input});

  },

  removeLeadsFromList: function(listId, leadIds) {
    if (!_.isArray(leadIds) && !_.isEmpty(leadIds)) {
      var msg = 'input must be an array of lead ids';
      log.error(msg);
      return Promise.reject(msg);
    }

    var path = util.createPath('lists', listId, 'leads.json');
    leadIds = [].concat(leadIds); // ensure it's an array
    return this._connection.del(path + "?id=" + leadIds.join('&id='), {headers: {"Content-Type": "application/json"}});
  },

  addEmailsToList: function(emails, listId, options) {
    options = options || {};

    emails = _.map(emails, function(email) {
      return {email: email};
    });

    return this._marketo.lead.createOrUpdate(emails, {lookupField: 'email'})
      .then(_.bind(function(data) {
        return this.addLeadsToList(listId, data.result);
      }, this));
  },

  getLeads: function(listId, options) {
    var path = util.createPath('list', listId, 'leads.json');
    options = _.extend({}, options, {
      listId: listId,
      _method: 'GET'
    });
    return this._connection.post(path, {data: options});
  },

  isMember: function(listId, leadIds) {
    var path = util.createPath('lists', listId, 'leads', 'ismember.json');
    leadIds = [].concat(leadIds); // ensure it's an array
    return this._connection.get(path + "?id=" + leadIds.join('&id='));
  },

  // convenience function to find a single list by id
  byId: function(listId) {
    return this.find({id: [listId]});
  }
};

module.exports = List;

File: src.net/Marketo-Rest/Api/List.cs
using System.Threading.Tasks;
using System.Linq;

namespace Marketo.Api
{
    /// <summary>
    /// Class List.
    /// </summary>
    public class List
    {
        readonly MarketoClient _marketo;
        readonly Connection _connection;

        /// <summary>
        /// Initializes a new instance of the <see cref="List"/> class.
        /// </summary>
        /// <param name="marketo">The marketo.</param>
        /// <param name="connection">The connection.</param>
        public List(MarketoClient marketo, Connection connection)
        {
            _marketo = marketo;
            _connection = connection;
        }

        /// <summary>
        /// Finds the specified options.
        /// </summary>
        /// <param name="options">The options.</param>
        /// <returns>Task&lt;dynamic&gt;.</returns>
        public Task<dynamic> Find(dynamic options = null)
        {
            options = dyn.exp(options);
            options = util.arrayToCSV(options, new[] { "id", "name", "programName", "workspaceName" });
            var path = util.createPath("lists.json");
            return _connection.get(path, new { data = options });
        }

        /// <summary>
        /// Adds the leads to list.
        /// </summary>
        /// <param name="listId">The list identifier.</param>
        /// <param name="input">The input.</param>
        /// <param name="options">The options.</param>
        /// <returns>Task&lt;dynamic&gt;.</returns>
        public Task<dynamic> AddLeadsToList(int listId, object[] input, dynamic options = null)
        {
            var path = util.createPath("lists", listId.ToString(), "leads.json");
            input = input.Select(x =>
            {
                if (x is string || x is int) return (object)new { id = x.ToString() };
                var id = dyn.getProp<string>(x, "id") ?? dyn.getProp<int>(x, "id").ToString();
                return id != "0" ? (new { id }) : null;
            }).Where(x => x != null).ToArray();
            return _connection.postJson(path, new { input }, options);
        }

        /// <summary>
        /// Removes the leads from list.
        /// </summary>
        /// <param name="listId">The list identifier.</param>
        /// <param name="leadIds">The lead ids.</param>
        /// <param name="options">The options.</param>
        /// <returns>Task&lt;dynamic&gt;.</returns>
        public Task<dynamic> RemoveLeadsFromList(int listId, int[] leadIds, dynamic options = null)
        {
            var path = util.createPath("lists", listId.ToString(), "leads.json");
            return _connection.del(path, new { query = new { id = string.Join("&id=", leadIds) } }, "application/json");
        }

        /// <summary>
        /// Adds the emails to list.
        /// </summary>
        /// <param name="emails">The emails.</param>
        /// <param name="listId">The list identifier.</param>
        /// <param name="options">The options.</param>
        /// <returns>Task&lt;dynamic&gt;.</returns>
        public async Task<dynamic> AddEmailsToList(string[] emails, int listId, dynamic options = null)
        {
            var emailsAsObj = emails.Select(email => new { email }).ToArray();
            var data = await _marketo.Lead.CreateOrUpdate(emailsAsObj, new { lookupField = "email" });
            data = await AddLeadsToList(listId, (object[])data.result, options);
            return data;
        }

        /// <summary>
        /// Gets the leads.
        /// </summary>
        /// <param name="listId">The list identifier.</param>
        /// <param name="options">The options.</param>
        /// <returns>Task&lt;dynamic&gt;.</returns>
        public Task<dynamic> GetLeads(int listId, dynamic options = null)
        {
            var path = util.createPath("list", listId.ToString(), "leads.json");
            options = dyn.exp(options);
            options.listId = listId;
            return _connection.post(path, new { data = options });
        }

        /// <summary>
        /// Determines whether the specified list identifier is member.
        /// </summary>
        /// <param name="listId">The list identifier.</param>
        /// <param name="leadIds">The lead ids.</param>
        /// <param name="options">The options.</param>
        /// <returns>Task&lt;dynamic&gt;.</returns>
        public Task<dynamic> IsMember(int listId, int[] leadIds, dynamic options = null)
        {
            var path = util.createPath("lists", listId.ToString(), "leads", "ismember.json");
            return _connection.get(path, new { query = new { id = string.Join("&id=", leadIds) } });
        }

        /// <summary>
        /// Bies the identifier.
        /// </summary>
        /// <param name="listId">The list identifier.</param>
        /// <returns>Task&lt;dynamic&gt;.</returns>
        public Task<dynamic> ById(int listId)
        {
            return Find(new { id = new[] { listId } });
        }
    }
}
